### ISSUE TITLE
Considering Kaggle as host

### ISSUE THREAD

**Issue by norvig** [aimacode/aima-exercises#21]
The team at Kaggle are interested in exploring the possibility that they could host the exercises. There's a quick demo here: https://www.kaggle.com/moconnell/classroom-example-for-peter

Kaggle offers:

- A nice setup for professors to create groups for their class.
- Discussion forums.
- Students would fork a [notebook](https://www.kaggle.com/dansbecker/bayesian-learning) of exercises, and answer them.
- Students can set visibility to public, class-wide, or private.
- TAs/graders (or peers) could then review/grade the students answers.
- Of course, Kaggle has the ability to run programming contests. This could be either within a professor's class, or we could run contests open to anyone.

If we use this we still have a bunch of the same questions not yet answered:
- Do we create/edit exercises individually, or a notebook at a time?
- What is the process for curating new suggested exercises?
- What  is the mechanism for a canonical, permanent symbol name for an exercise?
- Can other hosts besides Kaggle also participate if they want to? If so, what is the process for keeping them in sync?

**Comment by Nalinc** [comment]
I agree that Kaggle supports some out-of-the-box features which otherwise would be tedious to implement. **Ability to fork individual notebooks** and support for **graders/peer-reviews** are some of the really cool features. I believe we should look more into it to see how many of these unanswered questions can be addressed. It is definitely worth exploring :)

**Comment by yakout** [comment]
@norvig I think Kaggle is very great choice. and here is my little suggestions:
> Do we create/edit exercises individually, or a notebook at a time?

Adding exercises per chapter will be more convenient. (i.e for notebooks we generated, each will be in a separate kernel)

> What is the process for curating new suggested exercises?

Adding or receiving a new pull request with new exercise in _aima-exercises_ repo, we must also update it’s corresponding kernel in Kaggle, that’s why I suggested that each file will be in a separate kernel, to easy maintain.

> What is the mechanism for a canonical, permanent symbol name for an exercise?

This really depends on how we want to organize it:
* by difficulty
* by it's appearance order in it's chapter
* addition order
* randomly

The last two options will be very easy to maintain, (Just incrementing the exercise number), also we must not contradict with the naming in the book.

> Can other hosts besides Kaggle also participate if they want to? If so, what is the process for keeping them in sync?

If I understood this point well, I think syncing will be using Github repo as well, updating one notebook in the repo, we must update it's corresponding kernel in Kaggle.

---

**Summary**: I think we should use both Jekyll and Kaggle.

**Jekyll with Github pages**; for making it open source repo that contains all the exercises in .md and .ipynb format, and receives pull requests for other new exercises created by professors or by others, or any other type of open contribution.

**Kaggle**; for the discussion forums, forking the exercises to add answers that will be graded or executed, plus other features you mentioned.

Looking forward for other suggestions as well.

**Comment by nvinayvarma189** [comment]
I agree with @Nalinc and @yakout here. I had a similar Idea about the project. We need a standard GitHub page (backed up with Jekyll or by any other means which supports MathJax) to display the exercises. The students who just want to see the exercises can refer to this page. 

Whereas for the students who want to get involved by creating new exercises, ranking exercises or projecting their views about any misconception regarding a question, we need to host a community. For this kaggle may be perfect. Then all the newly added exercises will come in as a Pull request and can be viewed in the standard page we maintain.

Again we need to keep a look for any issues that might occur in the future. Also, the limitations of hosting a community in kaggle must not affect our progress in implementing new ideas in the coming years. If we think so, We need to develop a community from scratch(pretty much difficult). I personally think that kaggle is pretty much flexible. If I'm missing something here, I kindly request everyone to let me know.

**Comment by Nalinc** [comment]
> Do we create/edit exercises individually, or a notebook at a time?
> > Adding exercises per chapter will be more convenient. (i.e for notebooks we generated, each will be in a separate kernel)

I agree that adding exercises per chapter will be more convenient. That's because we already have Jupyter notebooks and they can directly be imported to Kaggle. However, this is only convenient from an instructor's point of view, or for situations where instructors give assignments to the class. From a student's perspective, this essentially means that they will only be able to have discussions on kernel level. This could be a blocking factor because students most often find only certain questions in a chapter more challenging than others and want to discuss possible approaches only for those questions. Instructors might as well prefer addressing each question individually and provide solutions to only certain questions.

> What is the process for curating new suggested exercises?
> > Adding or receiving a new pull request with new exercise in aima-exercises repo, we must also update it’s corresponding kernel in Kaggle, that’s why I suggested that each file will be in a separate kernel, to easy maintain.

This basically means that one has to first fork the aima-exercises repo on Github, and once changes are merged, both(pull requester and merger) would have to sync their corresponding kernel in Kaggle (assuming the pull-requester has also forked the Kernel on Kaggle). Even if the exercises are divided on kernel level, the process is tedious because there's manual syncing involved on multiple ends.
 
To summarize, though Kaggle is good to host assignments and review and grade them, I don't think it is really promising as a discussion forum. Long comment threads on kernel level would not only be less attractive for students, but also difficult to track.

**Comment by KaustabhGanguly** [comment]
I agree with everything above . But I have a question , that why can't we use a full website with database integration ?

**Comment by heisenbuug** [comment]
I think Kaggle and notebooks combined  will be a good option.
